### PR TITLE
Duration type is a builtin type

### PR DESCRIPTION
**Pull request by Akirathan**
### Pull Request Description

- Reimplement the `Duration` type to a built-in type.
- `Duration` is an interop type.
- Allow Enso method dispatch on `Duration` interop coming from different languages.

### Important Notes

- The older `Duration` type should now be split into new `Duration` builtin type and a `Period` type.
  - This PR does not implement `Period` type, so all the `Period`-related functionality is currently not working, e.g., `Date - Period`.
- This PR removes `Integer.milliseconds`, `Integer.seconds`, ..., `Integer.years` extension methods.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

### CODE REVIEW

**Comment by Akirathan** [comment]
According to the associated issue description, we want to split the current functionality of `Duration` into `Duration` and `Period`, so I removed all the date-related functionality from `Duration`, and left only time-related functionality. All the removed date-related functionality should be moved to a new `Period` (maybe existing `Time_Period`?) builtin type in the future.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDateTime.java
Comment (hubertp) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDateTime.java]: You want to avoid using the uncached one. 
You can auto-generate the correct one by adding  `@Builtin.Specialize` and changing the definition to `public EnsoDateTime plus(Object durationObject, InteropLibrary interop) {`

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDateTime.java
Comment (hubertp) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDateTime.java]: Same comment as for `plus`

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDuration.java
Comment (hubertp) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDuration.java]: You don't want that since you added `EnsoDuration` to the list of accepted types. Need to update the annotation processor in `TypeWithKind` so that it is happy 

```suggestion
```

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDuration.java
Comment (hubertp) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDuration.java]:
```suggestion
```
same as for create and others

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDuration.java
Comment (hubertp) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDuration.java]: Similar comment about `interop` as for `plus` and `minus`

File: test/Tests/src/Semantic/Java_Interop_Spec.enso
Comment (hubertp) [inline test/Tests/src/Semantic/Java_Interop_Spec.enso]: I think you might want to add those to Duration_Spec, similarly as we have date-specific tests in `Date_Spec`, `Date_Time_Spec` etc.

**Review by hubertp — COMMENTED** [review]
The main implementation looks fine, but there are few cleanups necessary

File: engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java:346]: This is exact copy of date, time & co. handling. However in case of _array_ handling I decided to not do any wrapping, just let the `hasArrayElements` object flow thru the engine. Pavel, what do you think is better from Truffle perspective: wrap or let the original `TruffleObject` flow?

If _"let the original object flow"_, can you try it with duration and other `doConvertXYZ` methods as well?

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDateTime.java
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDateTime.java:10]: Replacing `*` imports with explicit once improves accuracy. I am all for it. However we should try to avoid changing imports just because our IDEs have explicit opinions on their particular order...

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDuration.java
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDuration.java]: `TruffleObject` cannot be subclassed as far as I know - shall we make these classes `final`?

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/type/Types.java
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/type/Types.java:55]: The ultimate goal is to get rid of all `TruffleObject` classes in this `Types` list. Technically nothing happens if we remove all date&time related ones. The important once to remove are however `Array` and `Vector` - tracked as [Array vs. ArrayOverBuffer](https://www.pivotaltracker.com/n/projects/2539304/stories/183058716) and [many builtins](https://www.pivotaltracker.com/n/projects/2539304/stories/183266964).

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDateTime.java
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDateTime.java:157]: **Request**: Changing builtins will very likely need changes to reflective files like in #3755 - please run `sbt engine-runner-native/buildNativeImage` to verify it builds properly. CCing @hubertp , @mwu-tow .

File: std-bits/base/src/main/java/org/enso/base/Time_Utils.java
Comment (JaroslavTulach) [inline std-bits/base/src/main/java/org/enso/base/Time_Utils.java:83]: Surprising. How do you handle `Period`? It is not Truffle interop type and needs to be still treated via host interop, I think.

File: test/Tests/src/Data/Time/Duration_Spec.enso
Comment (JaroslavTulach) [inline test/Tests/src/Data/Time/Duration_Spec.enso:71]: **Request** (_at least for explanation_): Removing these tests seems suspicious to me. I believe `month` or `year` is a case that requires `Period` and not `Duration` - does it mean `Period` will no longer be supported, if this PR is integrated? CCing @radeusgd and @jdunkerley .

**Review by JaroslavTulach — APPROVED** [review]
The changes that introduce `EnsoDuration` seems fine to me. They directly copy what has been done previously and that is acceptable. Let few notes there related to some future changes...

I have a feeling this PR removes support for `Period`. Not sure whether that is OK or not? (marked as **request**)

Please manually check that native image still builds and can execute something simple. (marked as **request**)

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDateTime.java
Comment (radeusgd) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDateTime.java:10]: Shouldn't we just agree on an ordering that we want and try to enforce it?

Technically we are using the Google Java Style [(according to our docs)](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md#code-formatting), which defines that [imports should be sorted in ASCII order](https://google.github.io/styleguide/javaguide.html#s3.3.3-import-ordering-and-spacing) (more or less).

Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDateTime.java:10]: I care mostly about not using `*` import. Anyway until we resolve [Java formatter doesn't work on files that use latest syntax](https://www.pivotaltracker.com/story/show/182515732) - we are probably left to _human judgment of common sense_.

Comment (radeusgd) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDateTime.java:10]: Btw. I've downloaded the formatter as a standalone JAR and downloaded Java 17 using SDKman and sometimes if I want to format a file using new syntax, I do:
```
/home/radeusgd/.sdkman/candidates/java/current/bin/java -jar ~/Downloads/google-java-format-1.15.0-all-deps.jar --replace path/to/java/file.java
```

and that works flawlessly :)

Of course it's a pity it's not automated, but it's some middle ground.

**Comment by JaroslavTulach** [comment]
Thanks for making the `TruffleObject` implementations `final`, Pavel.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDateTime.java
Comment (hubertp) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDateTime.java:157]: To maybe complement @JaroslavTulach's comment - once it finishes (it will take a while) you should be able to execute `./runner --run engine/runner-native/src/test/resources/Factorial.enso 6` without errors

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Time_Of_Day.enso
Comment (hubertp) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Time_Of_Day.enso:220]: Any particular reason why we should prefer builtins rather than calling Java code? This is a general question, also to @JaroslavTulach

File: test/Tests/src/Data/Time/Date_Spec.enso
Comment (hubertp) [inline test/Tests/src/Data/Time/Date_Spec.enso]: And this passes now? Call me surprised :)

File: test/Tests/src/Data/Time/Duration_Spec.enso
Comment (hubertp) [reply test/Tests/src/Data/Time/Duration_Spec.enso:71]: I believe they've been just moved around, right @Akirathan ?

File: test/Tests/src/Semantic/Java_Interop_Spec.enso
Comment (hubertp) [inline test/Tests/src/Semantic/Java_Interop_Spec.enso]:
```suggestion
```
spurious (now)

File: test/Tests/src/Semantic/Java_Interop_Spec.enso
Comment (hubertp) [inline test/Tests/src/Semantic/Java_Interop_Spec.enso]:
```suggestion
```

**Review by hubertp — APPROVED** [review]
LGTM. Thanks for making the changes

File: CHANGELOG.md
Comment (radeusgd) [inline CHANGELOG.md]:
```suggestion
- [Reimplemented `Duration` as a built-in type.][3759]
```

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Duration.enso
Comment (radeusgd) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Duration.enso]: I think this can be removed no, right?

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Date.enso
Comment (radeusgd) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Date.enso]:
```suggestion
from Standard.Base.Error.Common import Time_Error_Data, unimplemented
```

We don't have a style guide and I guess both approaches are OK, I'm wondering if we should keep the imports from a single module in one line? I think most of the time that's what we've been doing.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDuration.java
Comment (radeusgd) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDuration.java]: I think we should handle/wrap the `ArithmeticException`. Here and also in `milliseconds`.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Duration.enso
Comment (radeusgd) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Duration.enso:172]: If I understand correctly, we change the semantics here.

Earlier this was returning the nanos part of the duration, i.e. if I have a duration of 1s 300ms, it would return `300`, now it will return `1300`. That's why I added the separate `total_milliseconds` method.

Is the semantics change deliberate or accidental?

Still, even if we want to keep the changed semantics, we definitely must update the docs to reflect that - `Get the portion of the duration expressed in milliseconds.` suggests the old semantics!

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Duration.enso
Comment (radeusgd) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Duration.enso]: I don't understand this change.

Why would I ever need a vector in which the same duration is expressed in various units?

The original meaning of this function was to represent the duration by each part so that the whole vector would reflect the whole duration. Now each element reflects the whole duration (with some partial precision). Do we need such a method?

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Duration.enso
Comment (radeusgd) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Duration.enso:183]: Earlier this method would return the seconds _part_ so it made sense for it to be an integer - a part will always be an integer.

Now if we are returning the whole duration in seconds, if we want to keep this new semantics, I think we should reconsider making it return a `Decimal` instead so the millisecond part is not truncated. If it is to be truncated, we must mention in the documentation that the milliseconds get truncated and specify what method of rounding is used.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDateTime.java
Comment (radeusgd) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDateTime.java:154]: Normally if a type mismatch happens, we will throw a `Type_Error` or `Unsupported_Argument_Types`.

If I understand this code correctly, here instead we will throw `UnsupportedMessageException`. I think maybe we should wrap it into some Enso-specific type? I think we could re-use one of the two above or alternatively create a new one. I'm just not convinced we should expose a Truffle exception type to our users - instead we should have an exception type that shows up somewhere in Enso like the two above (i.e. it has its `type Type_Error` definition with documentation, `to_display_text` etc.)

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDuration.java
Comment (radeusgd) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDuration.java]: Do I understand correctly that I can do
`Duration.between (Date.new 1999 1 2) (Date_Time.new 1999 4 8 12 40)` and with `convertToDateTime` the former will be converted to a Date Time @ `1999-01-02 0:00:00` and happily compute the difference?

I think passing a Date/Date_Time pair should result in some Type_Error or Illegal_Argument_Error - we should either have two Dates or two Date_Times, not a mix of them. The mix is going to likely be a user error - if the user really wants to compute a difference between mixed types they can perform a conversion explicitly.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDuration.java
Comment (radeusgd) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDuration.java]: This docstring (not sure how important it is and where it shows up?) is a bit unclear. Is this the total duration expressed in milliseconds or the millisecond _part_ of the duration? (I know it is the former after reading the code carefully, but the doc should IMO answer that question without requiring to read the code - otherwise why have the doc at all)

File: test/Tests/src/Data/Time/Duration_Spec.enso
Comment (radeusgd) [inline test/Tests/src/Data/Time/Duration_Spec.enso]: So where is the test for other units?

**Review by radeusgd — COMMENTED** [review]
The code looks good, but as mentioned in the in-line comments, the semantics of the `seconds`, `milliseconds` etc. getters has been altered.

I want to confirm if this change was deliberate, as I'm not aware of any discussions to change this part of the design. IMO the old way where you could get each part was useful - and we could also have the `total_` variants of each to get the full duration expressed in a given unit - giving us both capabilities.

Still, if it was deliberate and we want to keep the new semantics, the documentation needs to be updated, and `milliseconds` and `nanoseconds` getters should handle the `ArithmeticException`, converting it to some dataflow error (maybe `Illegal_State_Error`?).

And we should revise if we want the `to_vector` method on Duration - with the changed semantics it does not seem useful at all.

**Comment by Akirathan** [comment]
As discussed with @radeusgd, I reverted the incompatible changes that I introduced in this PR. So the only breaking changes are:
* removing all `Integer` extension methods, as we want to explicitly differentiate between `Duration` and `Period`.
* Ignore all tests that are related to `Period`, i.e., that will be failing untill `Period` is implemented.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Time_Of_Day.enso
Comment (JaroslavTulach) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Time_Of_Day.enso:220]: Builtins are necessary when we need to deal with `TruffleObject` and use a bit of partial evaluation. Otherwise it is better if the libraries stick with classical (hosted) Java interop.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDateTime.java
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDateTime.java:10]: We'd have to put that into CI check to enforce the fomatting...

**Review by JaroslavTulach — APPROVED** [review]

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDateTime.java
Comment (radeusgd) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDateTime.java:10]: I think we could implement such a check running on some default Java 19 JVM.

Still, until we do, I think we can just agree to try to keep the files formatted using this method, we will miss some, but I think that's better than not doing it at all.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Date_Time.enso
Comment (radeusgd) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Date_Time.enso]: I don't think this should be wrapped in `Polyglot_Error_Data`. This is a wrapper in which Java exceptions are usually wrapped - so the field `cause` is supposed to be the exception, not a `Text` message. And we have never (IIRC) used `Polyglot_Error` as a dataflow error - it's a marker for Java exceptions coming from builtin/polyglot calls _and_ there are plans to remove it too. So I'm against using it in this context.

Please use some more fitting exception. Maybe `Illegal_State_Error`? Or maybe `Time_Error` - I think it is fitting what we need here.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Duration.enso
Comment (radeusgd) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Duration.enso]: minor nitpick: this function has no arguments (okay, other than `self`), so maybe `Illegal_State_Error`? Because the result is independent of arguments - it depends on the internal _state_ of `self`.

But feel free to leave as-is too, I don't think that's very important. Maybe we should decide on some robust guideline when to use which.

File: distribution/lib/Standard/Test/0.0.0-dev/src/Main.enso
Comment (radeusgd) [inline distribution/lib/Standard/Test/0.0.0-dev/src/Main.enso]: I'm wondering if it wouldn't be more readable if we made a `Duration.zero` alias which calls `new` with no arguments. Would be clearer that it is a 0-duration.

**Review by radeusgd — COMMENTED** [review]
Please change the `Polyglot_Error` before merging, otherwise gtg.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDuration.java
Comment (radeusgd) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDuration.java]: This is an internal error that should never happen, right? Because we just assert that `isDate && isTime`.

Then I think the convention is to throw `IllegalStateException`, see for example `ContextRewrapNode.java:113`.

And other than that for future - the I think `throw new PanicException(x)` essentially acts as `Panic.throw x` in Enso.
So what you are doing is `Panic.throw e.getMessage` - i.e. throwing a pure `Text` value. We actually used to do that a lot, but have since moved to throwing meaningful atoms - because if you throw just Text it's hard to catch it - you can just `Panic.catch Text` but you cannot do more filtering easily, whereas we have facilities for catching a specific exception type. So in general it would be good to put some atom as the payload for such panics (the atom itself may for example contain a varying message, but it should be some type so that we can distinguish types of exceptions).

**Review by radeusgd — APPROVED** [review]

**Pull request merged**
